<a href="https://colab.research.google.com/github/Madihasafi/Midterm-Datavis/blob/main/Midterm_Datavis_Code.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Analyze flight delay data from the "Flight_Delays_Complete_Analysis_2003-2025_Detailed.csv" file to determine airline punctuality across years and airport punctuality. Generate interactive Plotly visualizations to identify the most punctual years for airlines and the most/least punctual airports. The final output should include the Python code for the analysis and visualizations, a summary of key insights, and explanations of design decisions.

## Load New Dataset

### Subtask:
Load the flight delay data from the CSV file 'Flight_Delays_Complete_Analysis_2003-2025_Detailed.csv' into a pandas DataFrame. This will use a larger, more comprehensive dataset for analysis.


**Reasoning**:
Load the flight delay data from 'Flight_Delays_Complete_Analysis_2003-2025_Detailed.csv' into a pandas DataFrame, display its head, and print its information to inspect its structure and data types.



In [11]:
df = pd.read_csv('Flight_Delays_Complete_Analysis_2003-2025_Detailed.csv')

print("First 5 rows of the loaded DataFrame:")
print(df.head())

print("\nDataFrame Information:")
df.info()

First 5 rows of the loaded DataFrame:
   Rank                                            Airport  \
0     1       Daniel K. Inouye International Airport (HNL)   
1     2                Orlando International Airport (MCO)   
2     3  Fort Lauderdale-Hollywood International Airpor...   
3     4                Oakland International Airport (OAK)   
4     5                              Kahului Airport (OGG)   

   Security delays per 1,000 delayed flights  \
0                                        7.0   
1                                        6.3   
2                                        5.5   
3                                        4.7   
4                                        4.7   

   Average delay length (>14 minutes)  Unnamed: 4  Unnamed: 5  Unnamed: 6  \
0                                34.3         NaN         NaN         NaN   
1                                45.1         NaN         NaN         NaN   
2                                31.9         NaN         NaN        

## Initial Data Cleaning and Feature Engineering

### Subtask:
Perform initial data cleaning on the new dataset, including handling missing values, standardizing column names, and extracting airport codes for analysis. Note that due to the structure of the loaded dataset, it is not possible to create an 'is_punctual' column based on individual flight arrival delays as originally intended.


**Reasoning**:
To perform the initial data cleaning as instructed, I will drop columns with all NaN values, rename specific columns for clarity, and extract the three-letter airport codes into a new column, then display the cleaned DataFrame's head and information to confirm the changes.



In [12]:
import re

# 1. Identify and drop columns with all NaN values and the descriptive column
# Get columns that are entirely NaN
columns_to_drop = df.columns[df.isnull().all()].tolist()

# Also explicitly drop the long descriptive column if it exists
long_desc_col = "Austin-Bergstrom has more security delays than most US airports, new study finds\nAustin-Bergstrom had the 12th-most delays related to security."
if long_desc_col in df.columns:
    columns_to_drop.append(long_desc_col)

df = df.drop(columns=columns_to_drop)

# 2. Rename the column 'Security delays per 1,000 delayed flights' to 'SecurityDelaysPer1000Flights'
df.rename(columns={'Security delays per 1,000 delayed flights': 'SecurityDelaysPer1000Flights'}, inplace=True)

# 3. Rename the column 'Average delay length (>14 minutes)' to 'AverageDelayLengthMinutes'
df.rename(columns={'Average delay length (>14 minutes)': 'AverageDelayLengthMinutes'}, inplace=True)

# 4. Extract the three-letter airport code from the 'Airport' column
# The regex pattern looks for three word characters inside parentheses
df['AirportCode'] = df['Airport'].str.extract(r'\((\w{3})\)')

# 5. Display the first 5 rows of the cleaned DataFrame and print its information
print("First 5 rows of the cleaned DataFrame:")
print(df.head())

print("\nCleaned DataFrame Information:")
df.info()

First 5 rows of the cleaned DataFrame:
   Rank                                            Airport  \
0     1       Daniel K. Inouye International Airport (HNL)   
1     2                Orlando International Airport (MCO)   
2     3  Fort Lauderdale-Hollywood International Airpor...   
3     4                Oakland International Airport (OAK)   
4     5                              Kahului Airport (OGG)   

   SecurityDelaysPer1000Flights  AverageDelayLengthMinutes AirportCode  
0                           7.0                       34.3         HNL  
1                           6.3                       45.1         MCO  
2                           5.5                       31.9         FLL  
3                           4.7                       44.6         OAK  
4                           4.7                       32.2         OGG  

Cleaned DataFrame Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 5 columns):
 #   Column     

## Analyze Airline Punctuality by Year

### Subtask:
Calculate the punctuality rate for each airline across different years from the new dataset to identify their most punctual years.


## Analyze Airline Punctuality by Year

### Subtask:
Calculate the punctuality rate for each airline across different years from the new dataset to identify their most punctual years.

#### Instructions
1. Verify the columns available in the current `df` by printing `df.columns`.
2. Observe that the `df` loaded from 'Flight_Delays_Complete_Analysis_2003-2025_Detailed.csv' does not contain the 'Year', 'Airline', or 'is_punctual' columns that would be necessary to perform the analysis of airline punctuality across years as outlined in this subtask.
3. Conclude that this specific subtask cannot be completed with the current dataset.


**Reasoning**:
To follow the first instruction of the subtask, I will print the columns of the `df` DataFrame to verify which columns are available for analysis.



In [13]:
print("Columns in the DataFrame:")
print(df.columns)

Columns in the DataFrame:
Index(['Rank', 'Airport', 'SecurityDelaysPer1000Flights',
       'AverageDelayLengthMinutes', 'AirportCode'],
      dtype='object')


## Analyze Airport Punctuality

### Subtask:
Determine the punctuality rates for airports from the new dataset to identify the most and least punctual airports, using available delay metrics.


**Reasoning**:
Since the direct punctuality rate calculation is not feasible with the current dataset, I will proceed by identifying the most and least punctual airports based on the available metrics: 'SecurityDelaysPer1000Flights' and 'AverageDelayLengthMinutes', as specified in the instructions. This will involve sorting the DataFrame by each of these columns in both ascending and descending order and displaying the top 5 relevant entries.



In [14]:
print("\n--- Airport Punctuality Analysis ---")

# 3. Identify the 5 most punctual airports based on 'SecurityDelaysPer1000Flights'
print("\nTop 5 Most Punctual Airports by Security Delays (Lower is better):")
most_punctual_security = df.sort_values(by='SecurityDelaysPer1000Flights', ascending=True)
print(most_punctual_security[['Airport', 'AirportCode', 'SecurityDelaysPer1000Flights']].head(5))

# 4. Identify the 5 least punctual airports based on 'SecurityDelaysPer1000Flights'
print("\nTop 5 Least Punctual Airports by Security Delays (Higher is worse):")
least_punctual_security = df.sort_values(by='SecurityDelaysPer1000Flights', ascending=False)
print(least_punctual_security[['Airport', 'AirportCode', 'SecurityDelaysPer1000Flights']].head(5))

# 5. Identify the 5 most punctual airports based on 'AverageDelayLengthMinutes'
print("\nTop 5 Most Punctual Airports by Average Delay Length (Lower is better):")
most_punctual_avg_delay = df.sort_values(by='AverageDelayLengthMinutes', ascending=True)
print(most_punctual_avg_delay[['Airport', 'AirportCode', 'AverageDelayLengthMinutes']].head(5))

# 6. Identify the 5 least punctual airports based on 'AverageDelayLengthMinutes'
print("\nTop 5 Least Punctual Airports by Average Delay Length (Higher is worse):")
least_punctual_avg_delay = df.sort_values(by='AverageDelayLengthMinutes', ascending=False)
print(least_punctual_avg_delay[['Airport', 'AirportCode', 'AverageDelayLengthMinutes']].head(5))


--- Airport Punctuality Analysis ---

Top 5 Most Punctual Airports by Security Delays (Lower is better):
                                              Airport AirportCode  \
14                  Miami International Airport (MIA)         MIA   
13               Portland International Airport (PDX)         PDX   
12         Chicago Midway International Airport (MDW)         MDW   
11  Louis Armstrong New Orleans International Airp...         MSY   
10       Austin-Bergstrom International Airport (AUS)         AUS   

    SecurityDelaysPer1000Flights  
14                           2.5  
13                           2.6  
12                           2.8  
11                           2.9  
10                           3.0  

Top 5 Least Punctual Airports by Security Delays (Higher is worse):
                                             Airport AirportCode  \
0       Daniel K. Inouye International Airport (HNL)         HNL   
1                Orlando International Airport (MCO)         MCO

## Visualize Airline Punctuality

### Subtask:
Generate an interactive Plotly line chart to display airline punctuality trends across years using the processed data. The visualization should include appropriate labels and legends.


## Visualize Airline Punctuality

### Subtask:
Generate an interactive Plotly line chart to display airline punctuality trends across years using the processed data. The visualization should include appropriate labels and legends.

#### Instructions
1.  Recall the conclusion from the 'Analyze Airline Punctuality by Year' subtask, where it was determined that the current `df` DataFrame does not contain the necessary 'Year', 'Airline', or 'is_punctual' columns.
2.  Conclude that generating a Plotly line chart for airline punctuality trends across years is not feasible with the current dataset due to the absence of these critical columns.

## Visualize Airline Punctuality

### Subtask:
Generate an interactive Plotly line chart to display airline punctuality trends across years using the processed data. The visualization should include appropriate labels and legends.

#### Instructions
1.  Recall the conclusion from the 'Analyze Airline Punctuality by Year' subtask, where it was determined that the current `df` DataFrame does not contain the necessary 'Year', 'Airline', or 'is_punctual' columns.
2.  Conclude that generating a Plotly line chart for airline punctuality trends across years is not feasible with the current dataset due to the absence of these critical columns.

```markdown
## Visualize Airline Punctuality

### Subtask:
Generate an interactive Plotly line chart to display airline punctuality trends across years using the processed data. The visualization should include appropriate labels and legends.

#### Instructions
1.  Recall the conclusion from the 'Analyze Airline Punctuality by Year' subtask, where it was determined that the current `df` DataFrame does not contain the necessary 'Year', 'Airline', or 'is_punctual' columns.
2.  Conclude that generating a Plotly line chart for airline punctuality trends across years is not feasible with the current dataset due to the absence of these critical columns.
```

## Visualize Airline Punctuality

### Subtask:
Generate an interactive Plotly line chart to display airline punctuality trends across years using the processed data. The visualization should include appropriate labels and legends.

#### Instructions
1.  Recall the conclusion from the 'Analyze Airline Punctuality by Year' subtask, where it was determined that the current `df` DataFrame does not contain the necessary 'Year', 'Airline', or 'is_punctual' columns.
2.  Conclude that generating a Plotly line chart for airline punctuality trends across years is not feasible with the current dataset due to the absence of these critical columns.


## Visualize Airport Punctuality

### Subtask:
Generate interactive Plotly bar charts to show the punctuality rates of the most and least punctual origin and destination airports using the processed data. The visualization should include appropriate labels and legends.


**Reasoning**:
To visualize airport punctuality, I will create four interactive bar charts using `plotly.express`: one each for the most punctual by security delays, least punctual by security delays, most punctual by average delay length, and least punctual by average delay length, ensuring proper labels and titles for each.



In [15]:
import plotly.express as px

# Create a bar chart for most punctual airports by security delays
fig_most_security = px.bar(
    most_punctual_security.head(5),
    x='AirportCode',
    y='SecurityDelaysPer1000Flights',
    title='Top 5 Most Punctual Airports by Security Delays',
    labels={
        'AirportCode': 'Airport',
        'SecurityDelaysPer1000Flights': 'Security Delays per 1000 Flights'
    },
    color='SecurityDelaysPer1000Flights',
    color_continuous_scale=px.colors.sequential.Plotly3
)
fig_most_security.show()

# Create a bar chart for least punctual airports by security delays
fig_least_security = px.bar(
    least_punctual_security.head(5),
    x='AirportCode',
    y='SecurityDelaysPer1000Flights',
    title='Top 5 Least Punctual Airports by Security Delays',
    labels={
        'AirportCode': 'Airport',
        'SecurityDelaysPer1000Flights': 'Security Delays per 1000 Flights'
    },
    color='SecurityDelaysPer1000Flights',
    color_continuous_scale=px.colors.sequential.Reds
)
fig_least_security.show()

# Create a bar chart for most punctual airports by average delay length
fig_most_avg_delay = px.bar(
    most_punctual_avg_delay.head(5),
    x='AirportCode',
    y='AverageDelayLengthMinutes',
    title='Top 5 Most Punctual Airports by Average Delay Length',
    labels={
        'AirportCode': 'Airport',
        'AverageDelayLengthMinutes': 'Average Delay Length (Minutes)'
    },
    color='AverageDelayLengthMinutes',
    color_continuous_scale=px.colors.sequential.Plotly3
)
fig_most_avg_delay.show()

# Create a bar chart for least punctual airports by average delay length
fig_least_avg_delay = px.bar(
    least_punctual_avg_delay.head(5),
    x='AirportCode',
    y='AverageDelayLengthMinutes',
    title='Top 5 Least Punctual Airports by Average Delay Length',
    labels={
        'AirportCode': 'Airport',
        'AverageDelayLengthMinutes': 'Average Delay Length (Minutes)'
    },
    color='AverageDelayLengthMinutes',
    color_continuous_scale=px.colors.sequential.Reds
)
fig_least_avg_delay.show()

## Final Task

### Subtask:
Summarize the key insights obtained from the analysis of the larger dataset regarding airline and airport punctuality. Also, explain any design decisions made during the analysis and visualization process.


## Summary:

### Q&A
*   **Were the most punctual years for airlines identified?**
    No, this analysis could not be performed because the loaded dataset lacked the necessary 'Year', 'Airline', and 'is_punctual' columns.
*   **Were the most and least punctual airports identified?**
    Yes, the most and least punctual airports were identified based on two metrics: 'Security Delays per 1,000 delayed flights' and 'Average Delay Length (minutes)'.
*   **Were interactive Plotly visualizations generated to identify the most punctual years for airlines?**
    No, interactive Plotly visualizations for airline punctuality across years could not be generated due to the absence of crucial 'Year', 'Airline', and 'is_punctual' columns in the dataset.
*   **Were interactive Plotly visualizations generated to identify the most/least punctual airports?**
    Yes, interactive Plotly bar charts were successfully generated to visualize the most and least punctual airports based on 'Security Delays per 1,000 delayed flights' and 'Average Delay Length (minutes)'.

### Data Analysis Key Findings
*   The initial dataset, 'Flight_Delays_Complete_Analysis_2003-2025_Detailed.csv', contained 15 entries and 13 columns.
*   Upon inspection, 9 of the columns were entirely empty or contained descriptive text, leading to the conclusion that this dataset was likely a summary or metadata file rather than a comprehensive flight delay log.
*   Empty columns and a descriptive column ("Austin-Bergstrom has more security delays...") were successfully dropped.
*   Relevant columns were renamed for clarity: 'Security delays per 1,000 delayed flights' became 'SecurityDelaysPer1000Flights', and 'Average delay length (>14 minutes)' became 'AverageDelayLengthMinutes'.
*   Three-letter airport codes were successfully extracted into a new 'AirportCode' column from the 'Airport' name.
*   **Airline punctuality analysis by year was not feasible** due to the absence of 'Year', 'Airline', or 'is_punctual' columns in the processed DataFrame.
*   **Airport Punctuality by Security Delays per 1,000 Flights:**
    *   **Most Punctual:** Miami International Airport (MIA) with 2.5 security delays per 1,000 flights, followed by Portland International Airport (PDX) at 2.6.
    *   **Least Punctual:** Daniel K. Inouye International Airport (HNL) with 7.0 security delays per 1,000 flights, and Orlando International Airport (MCO) at 6.3.
*   **Airport Punctuality by Average Delay Length (minutes):**
    *   **Most Punctual:** Fort Lauderdale-Hollywood International Airport (FLL) with an average delay length of 31.9 minutes, followed by Kahului Airport (OGG) at 32.2 minutes.
    *   **Least Punctual:** Seattle-Tacoma International Airport (SEA) with the longest average delay length of 57.8 minutes, and Miami International Airport (MIA) at 55.4 minutes.
*   Interactive Plotly bar charts were successfully generated to visualize the top 5 most and least punctual airports for both security delays and average delay length.

### Insights or Next Steps
*   The current dataset is unsuitable for airline-specific or year-over-year punctuality analysis. To perform these analyses, a more detailed dataset containing individual flight records with airline identifiers, departure/arrival times, and delay reasons (if available) would be required.
*   The analysis successfully identified specific airports that excel or struggle in terms of security delays and average delay length. This information can be valuable for airport authorities or travelers, highlighting areas for potential improvement or consideration when planning travel.


# Task
Summarize the key insights obtained from the analysis of the flight delay data, reiterating the limitations due to the dataset's structure regarding year-based and weather-based analysis, and suggest what kind of data would be needed for such a comprehensive analysis.

## Load Flight Data

### Subtask:
Load the specified CSV file '/content/Flight_Delays_Complete_Analysis_2003-2025_Detailed (1).csv' into a pandas DataFrame to begin the analysis.


**Reasoning**:
To load the specified CSV file, I will use pandas' `read_csv` function, display its first 5 rows to ensure correct loading, and print its information to inspect its structure and data types as instructed.



In [16]:
import pandas as pd

# Load the CSV file into a pandas DataFrame
df = pd.read_csv('/content/Flight_Delays_Complete_Analysis_2003-2025_Detailed (1).csv')

# Display the first 5 rows of the DataFrame
print("First 5 rows of the loaded DataFrame:")
print(df.head())

# Print a summary of the DataFrame
print("\nDataFrame Information:")
df.info()

First 5 rows of the loaded DataFrame:
   Year Data_Period  Air_Carrier_Delay_Pct  Aircraft_Arriving_Late_Pct  \
0  2003     Jun-Dec                  30.64                       38.36   
1  2004   Full Year                  32.58                       37.04   
2  2005   Full Year                  31.88                       38.17   
3  2006   Full Year                  29.76                       40.33   
4  2007   Full Year                  30.42                       39.44   

   National_Aviation_System_Pct  Security_Delay_Pct  Extreme_Weather_Pct  \
0                         24.88                0.01                 6.11   
1                         23.49                0.00                 6.89   
2                         23.79                0.00                 6.16   
3                         24.34                0.00                 5.57   
4                         24.17                0.01                 5.96   

   Weather_Total_Share_Pct  Total_Delay_Minutes  Total_Delay

## Inspect and Clean Data

### Subtask:
Clean the data by dropping columns with mostly missing values and handling any remaining missing values.


**Reasoning**:
To clean the data as instructed, I will first identify and drop columns with a high number of missing values (specifically those with only 1 or very few non-null entries, and the 'Notes' column), then drop any rows that still contain missing values, and finally display the head and information of the resulting DataFrame to verify the cleaning process.



In [17]:
import numpy as np

# 1. Identify columns with a large number of missing values
# Based on df.info() from previous step, columns with only 1 non-null entry out of 22 are candidates.
# Also, 'Notes' column has only 4 non-null entries, making it mostly missing.
columns_to_drop_missing = [
    'Total_Delay_Minutes',
    'Total_Delay_Hours',
    'Air_Carrier_Hours',
    'Aircraft_Late_Hours',
    'NAS_Hours',
    'Weather_Hours',
    'Security_Hours',
    'Notes'
]

# 2. Drop these identified columns
df_cleaned = df.drop(columns=columns_to_drop_missing)

# 3. Check for any remaining missing values and drop rows if necessary
# 'Weather_Total_Share_Pct' has 1 missing value. We will drop rows with any missing values.
initial_rows = df_cleaned.shape[0]
df_cleaned.dropna(inplace=True)
final_rows = df_cleaned.shape[0]

print(f"Dropped {initial_rows - final_rows} rows due to remaining missing values.")

# 4. Display the first 5 rows of the cleaned DataFrame and print its information
print("\nFirst 5 rows of the cleaned DataFrame:")
print(df_cleaned.head())

print("\nCleaned DataFrame Information:")
df_cleaned.info()

Dropped 1 rows due to remaining missing values.

First 5 rows of the cleaned DataFrame:
   Year Data_Period  Air_Carrier_Delay_Pct  Aircraft_Arriving_Late_Pct  \
0  2003     Jun-Dec                  30.64                       38.36   
1  2004   Full Year                  32.58                       37.04   
2  2005   Full Year                  31.88                       38.17   
3  2006   Full Year                  29.76                       40.33   
4  2007   Full Year                  30.42                       39.44   

   National_Aviation_System_Pct  Security_Delay_Pct  Extreme_Weather_Pct  \
0                         24.88                0.01                 6.11   
1                         23.49                0.00                 6.89   
2                         23.79                0.00                 6.16   
3                         24.34                0.00                 5.57   
4                         24.17                0.01                 5.96   

   Weather

## Address Analysis Limitations

### Subtask:
Explicitly acknowledge that the dataset does not contain individual flight records, nor specific airport or airline data. Therefore, a direct analysis of 'airline punctuality' or 'airport punctuality' as originally requested (based on individual flight delays) is not possible with this dataset. The analysis will proceed with the available yearly delay percentage metrics.


**Reasoning**:
To explicitly acknowledge the limitations of the dataset, I will first display the columns of the `df_cleaned` DataFrame to show what data is available for analysis.



In [18]:
print("Columns in the cleaned DataFrame:")
print(df_cleaned.columns)

Columns in the cleaned DataFrame:
Index(['Year', 'Data_Period', 'Air_Carrier_Delay_Pct',
       'Aircraft_Arriving_Late_Pct', 'National_Aviation_System_Pct',
       'Security_Delay_Pct', 'Extreme_Weather_Pct', 'Weather_Total_Share_Pct'],
      dtype='object')


```markdown
Based on the available columns, it is clear that this dataset ('Flight_Delays_Complete_Analysis_2003-2025_Detailed (1).csv') does not contain individual flight records, nor specific airport or airline data in a granular format. Therefore, a direct analysis of 'airline punctuality' or 'airport punctuality' as originally requested (e.g., identifying punctuality of individual airlines or specific origin/destination airports based on individual flight delays) is not possible with this dataset.

Instead, this dataset is suitable for analyzing aggregated delay percentages across years for different categories, specifically:
- `Air_Carrier_Delay_Pct`: Percentage of delays attributable to the air carrier.
- `Aircraft_Arriving_Late_Pct`: Percentage of delays due to an earlier flight arriving late.
- `National_Aviation_System_Pct`: Percentage of delays due to National Aviation System impediments.
- `Security_Delay_Pct`: Percentage of delays due to security issues.
- `Extreme_Weather_Pct`: Percentage of delays due to extreme weather.
- `Weather_Total_Share_Pct`: Total share of weather-related delays.

Therefore, the analysis will proceed by focusing on the yearly trends and comparisons of these delay percentage categories, as this is what the current dataset supports.
```

## Analyze Yearly Delay Trends

### Subtask:
Identify yearly trends for different delay categories.


**Reasoning**:
To identify yearly trends for each delay category, I will iterate through the specified percentage columns, find the year(s) with the highest and lowest percentages for each, and then print these results, as specified in the instructions.



In [19]:
print("\n--- Yearly Delay Trends Analysis ---")

delay_columns = [
    'Air_Carrier_Delay_Pct',
    'Aircraft_Arriving_Late_Pct',
    'National_Aviation_System_Pct',
    'Security_Delay_Pct',
    'Extreme_Weather_Pct',
    'Weather_Total_Share_Pct'
]

for col in delay_columns:
    # Find year(s) with the highest percentage
    max_value = df_cleaned[col].max()
    max_years = df_cleaned[df_cleaned[col] == max_value]['Year'].tolist()
    print(f"\n{col}:")
    print(f"  Highest: {max_value:.2f}% in Year(s): {max_years}")

    # Find year(s) with the lowest percentage
    min_value = df_cleaned[col].min()
    min_years = df_cleaned[df_cleaned[col] == min_value]['Year'].tolist()
    print(f"  Lowest: {min_value:.2f}% in Year(s): {min_years}")


--- Yearly Delay Trends Analysis ---

Air_Carrier_Delay_Pct:
  Highest: 33.49% in Year(s): [2019]
  Lowest: 28.12% in Year(s): [2012]

Aircraft_Arriving_Late_Pct:
  Highest: 42.29% in Year(s): [2012]
  Lowest: 37.04% in Year(s): [2004]

National_Aviation_System_Pct:
  Highest: 24.88% in Year(s): [2003]
  Lowest: 23.10% in Year(s): [2010]

Security_Delay_Pct:
  Highest: 0.01% in Year(s): [2003, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019]
  Lowest: 0.00% in Year(s): [2004, 2005, 2006, 2020, 2021, 2022, 2023]

Extreme_Weather_Pct:
  Highest: 6.89% in Year(s): [2004]
  Lowest: 5.57% in Year(s): [2006]

Weather_Total_Share_Pct:
  Highest: 49.92% in Year(s): [2003]
  Lowest: 27.14% in Year(s): [2023]


## Visualize Yearly Delay Trends

### Subtask:
Generate interactive Plotly line charts to visualize the yearly trends for each delay category.


**Reasoning**:
To visualize the yearly trends for each delay category, I will import `plotly.express`, define a list of delay percentage columns, and then iterate through this list to generate and display an interactive Plotly line chart for each category, as specified in the instructions.



In [20]:
import plotly.express as px

# Create a list of the delay percentage columns to visualize
delay_columns = [
    'Air_Carrier_Delay_Pct',
    'Aircraft_Arriving_Late_Pct',
    'National_Aviation_System_Pct',
    'Security_Delay_Pct',
    'Extreme_Weather_Pct',
    'Weather_Total_Share_Pct'
]

# For each delay column, create and display an interactive line plot
for col in delay_columns:
    fig = px.line(
        df_cleaned,
        x='Year',
        y=col,
        title=f'Yearly Trend of {col.replace("_", " ")}',
        labels={
            'Year': 'Year',
            col: col.replace("_", " ")
        }
    )
    fig.show()

## Final Task

### Subtask:
Summarize the key insights obtained from the analysis of the flight delay data, reiterating the limitations due to the dataset's structure regarding year-based and weather-based analysis, and suggest what kind of data would be needed for such a comprehensive analysis.


## Summary:

### Data Analysis Key Findings

*   **Dataset Structure and Cleaning:** The initial dataset contained 22 entries and 16 columns. After cleaning, which involved dropping columns with substantial missing values (e.g., `Total_Delay_Minutes`, `Total_Delay_Hours`, `Notes`) and one row with a remaining missing value, the final DataFrame comprised 21 entries and 8 columns. The cleaned data exclusively contains yearly aggregated delay percentages rather than granular flight details.
*   **Analytical Scope Limitation:** Due to the aggregated nature of the data, the analysis could not address individual airline or airport punctuality, nor could it provide insights into specific flight-level delays. The analysis focused solely on yearly trends of aggregated delay categories.
*   **Key Yearly Delay Trends:**
    *   **Air Carrier Delays:** Peaked in 2019 at 33.49% and were lowest in 2012 at 28.12%.
    *   **Aircraft Arriving Late:** Was highest in 2012 at 42.29% and lowest in 2004 at 37.04%.
    *   **National Aviation System Delays:** Reached their highest in 2003 at 24.88% and lowest in 2010 at 23.10%.
    *   **Security Delays:** Consistently remained very low, with a highest percentage of 0.01% across several years (2003, 2007-2019) and 0.00% in others (2004-2006, 2020-2023), indicating a minimal contribution to overall delays.
    *   **Extreme Weather Delays:** Peaked in 2004 at 6.89% and were lowest in 2006 at 5.57%.
    *   **Total Weather Delays:** Showed a significant range, highest in 2003 at 49.92% and lowest in 2023 at 27.14%.
